Gems/Atom/Feature/Common/Code/Source/Mesh/ModelReloader.cpp

```diff
@@ -0,0 +1,173 @@
+                // Multiple meshes may reference the same buffer, so only add the dependency if it is unique
+                m_meshDependencies.push_back(asset);
+            }
+        }
+
+        void ModelReloader::ReloadDependenciesAndWait()
+        {
+            // Get the current list of dependencies depending on the current state
+            DependencyList& dependencies = GetPendingDependencies();
+
+            if (!m_pendingDependencies.none())
+            {
+                AZ_Assert(
+                    m_pendingDependencies.none(),
+                    "ModelReloader attempting to add new dependencies while still waiting for other dependencies in the hierarchy to "
+                    "load.");
+            }
+            if (dependencies.empty())
+            {
+                // If the original model asset failed to load, it won't have any dependencies to reload
+                AdvanceToNextLevelOfHierarchy();
+            }
+            AZ_Assert(dependencies.size() <= m_pendingDependencies.size(), "ModelReloader has more dependencies than can fit in the bitset. The size of m_pendingDependencies needs to be increased.");
+
+            // Set all bits to 1
+            m_pendingDependencies.set();
+            // Clear the least significant n-bits
+            m_pendingDependencies <<= dependencies.size();
+            // Set the least significant n-bits to 1, and the rest to 0
+            m_pendingDependencies.flip();
+
+            // Reload all the assets
+            for (Data::Asset<Data::AssetData>& dependencyAsset : dependencies)
+            {
+                Data::AssetBus::MultiHandler::BusConnect(dependencyAsset.GetId());
+                dependencyAsset.Reload();
+            }
+        }
+
+        void ModelReloader::AdvanceToNextLevelOfHierarchy()
+        {
+            switch (m_state)
+            {
+                case State::WaitingForMeshDependencies:
+                    m_state = State::WaitingForModelDependencies;
+                    ReloadDependenciesAndWait();
+                    break;
+                case State::WaitingForModelDependencies:
+                    m_state = State::WaitingForModel;
+                    ReloadDependenciesAndWait();
+                    break;
+                case State::WaitingForModel:
+                    Data::AssetBus::MultiHandler::BusDisconnect();
+                    // Since the model asset is finished reloading, orphan model from the instance database
+                    // so that all of the buffer instances are re-created with the latest data
+                    RPI::Model::TEMPOrphanFromDatabase(m_modelAsset.front());
+                    // Signal that the model is ready
+                    m_onModelReloaded.Signal(m_modelAsset.front());
+                    // Remove this reloader from the ModelReloaderSystem
+                    m_onRemoveReloaderFromSystem.Signal(m_modelAsset.front().GetId());
+                    delete this;
```

Comment:
Oh, here's that delete. I think this should be moved to the ModelReloaderSystem, that's more symmetrical, and it's pretty unexpected to have it here.

Reply:
This is a good suggestion, thanks for providing a solution.